statistics/table.go

```diff
@@ -68,6 +70,113 @@ type Table struct {
 	// and the schema of the table does not change, we don't need to load the stats for this
 	// table again.
 	TblInfoUpdateTS uint64
+
+	IsPkIsHandle bool
+}
+
+// ColAndIndexExistenceMap is the meta map for statistics.Table.
+// It can tell whether a column/index really has its statistics. So we won't send useless kv request when we do online stats loading.
+type ColAndIndexExistenceMap struct {
+	m1            map[int64]*model.ColumnInfo
+	colAnalyzed   map[int64]bool
+	m2            map[int64]*model.IndexInfo
+	indexAnalyzed map[int64]bool
+	someAnalyzed  bool
+}
+
+// SomeAnalyzed checks whether some part of the table is analyzed.
+// The newly added column/index might not be has its stats.
+func (m *ColAndIndexExistenceMap) SomeAnalyzed() bool {
+	if m == nil {
+		return false
+	}
+	for _, v := range m.colAnalyzed {
+		if v {
+			return true
+		}
+	}
+	for _, v := range m.indexAnalyzed {
+		if v {
+			return true
+		}
+	}
+	return false
+}
+
+// Has checks whether a column/index stats exists.
+func (m *ColAndIndexExistenceMap) Has(id int64, isIndex bool) bool {
+	if isIndex {
+		_, ok := m.m2[id]
+		return ok
+	}
+	_, ok := m.m1[id]
+	return ok
+}
+
+// HasAnalyzed checks whether a column/index stats exists and it has stats.
+// TODO: the map should only keep the analyzed cols.
+func (m *ColAndIndexExistenceMap) HasAnalyzed(id int64, isIndex bool) bool {
+	if isIndex {
+		analyzed, ok := m.indexAnalyzed[id]
+		return ok && analyzed
+	}
+	analyzed, ok := m.colAnalyzed[id]
+	return ok && analyzed
+}
+
+// InsertCol inserts a column with its meta into the map.
+func (m *ColAndIndexExistenceMap) InsertCol(id int64, info *model.ColumnInfo, analyzed bool) {
+	m.m1[id] = info
+	m.colAnalyzed[id] = analyzed
+}
+
+// GetCol gets the meta data of the given column.
+func (m *ColAndIndexExistenceMap) GetCol(id int64) *model.ColumnInfo {
+	return m.m1[id]
+}
+
+// InsertIndex inserts an index with its meta into the map.
+func (m *ColAndIndexExistenceMap) InsertIndex(id int64, info *model.IndexInfo, analyzed bool) {
+	m.m2[id] = info
+	m.indexAnalyzed[id] = analyzed
+}
+
+// GetIndex gets the meta data of the given index.
+func (m *ColAndIndexExistenceMap) GetIndex(id int64) *model.IndexInfo {
+	return m.m2[id]
+}
+
+// IsEmpty checks whether the map is empty.
+func (m *ColAndIndexExistenceMap) IsEmpty() bool {
+	return len(m.m1)+len(m.m2) == 0
+}
+
+// Clone deeply copies the map.
+func (m *ColAndIndexExistenceMap) Clone() *ColAndIndexExistenceMap {
+	mm := NewColAndIndexExistenceMap(len(m.m1), len(m.m2))
+	for k, v := range m.m1 {
```

Comment:
The name `m1` and `m2` seem a little unclear, can we rename them like `colInfo` and `indexInfo`?

Reply:
yes, I used those names at the beginning since there're just two maps. But later I found that we need to extend to four maps to make this pull as small as possible.